pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -84,17 +84,40 @@ public void init(List<TransformFunction> arguments, Map<String, DataSource> data
     }
     int numWhenStatements = arguments.size() / 2;
     _whenStatements = new ArrayList<>(numWhenStatements);
+    _elseThenStatements = new ArrayList<>(numWhenStatements + 1);
+    constructStatementList(arguments);
+    _selections = new boolean[_elseThenStatements.size()];
+    Collections.reverse(_elseThenStatements);
+    Collections.reverse(_whenStatements);
+    _resultMetadata = calculateResultMetadata();
+  }
+
+  private void constructStatementList(List<TransformFunction> arguments) {
+    int numWhenStatements = arguments.size() / 2;
+    boolean hasLegacyFormat = false;
+    for (int i = 0; i < numWhenStatements; i++) {
+      if (arguments.get(i * 2).getResultMetadata().getDataType() != DataType.BOOLEAN) {
+        hasLegacyFormat = true;
+        break;
```

Comment:
(MAJOR) This can break when the first loop didn't break but the second loop break because some wrong arguments already added to the list.
Also, we should check in favor of the new order. We can check if the first half of the arguments are all boolean, and if the odd index arguments are all boolean. If both are all booleans or both are not all booleans or the first half are all booleans, use the new order; use the legacy one otherwise.

Reply:
got it. yeah also this is not in critical path i will just make sure the check occurs afterward by looping it again

Reply:
`if (all_first_half_are_expression_boolean_type && odd_half_has_non_boolean_type) then use`